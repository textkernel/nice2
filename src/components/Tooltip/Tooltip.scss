.Tooltip {
    background-color: var(--color-neutral-90);
    color: var(--color-background);
    box-sizing: border-box;
    padding: var(--spacing-3x);
    border-radius: var(--border-radius);
    max-width: 350px;
    opacity: 0;
    animation: appearing 0.3s forwards;

    &__arrow {
        position: absolute;
    }

    [x-placement^="top"] & {
        margin-bottom: var(--spacing-5x);
        transform: translateY(-20px);
    }
    [x-placement^="top"] &__arrow {
        border-top: 6px solid var(--color-neutral-90);
        border-left: 6px solid transparent;
        border-right: 6px solid transparent;
        bottom: -6px;
        left: calc(50% - 6px);
        margin-top: 0;
        margin-bottom: 0;
    }
    [x-placement^="top-start"] &__arrow {
        left: var(--spacing-3x);
    }
    [x-placement^="top-end"] &__arrow {
        left: auto;
        right: var(--spacing-3x);
    }
    [x-placement^="bottom"] & {
        margin-top: var(--spacing-5x);
        transform: translateY(20px);
    }
    [x-placement^="bottom"] &__arrow {
        border-bottom: 6px solid var(--color-neutral-90);
        border-left: 6px solid transparent;
        border-right: 6px solid transparent;
        top: -6px;
        left: calc(50% - 6px);
        margin-top: 0;
        margin-bottom: 0;
    }
    [x-placement^="bottom-start"] &__arrow {
        left: var(--spacing-3x);
    }
    [x-placement^="bottom-end"] &__arrow {
        left: auto;
        right: var(--spacing-3x);
    }
    [x-placement^="right"] & {
        margin-left: var(--spacing-5x);
        transform: translateX(20px);
    }
    [x-placement^="right"] &__arrow {
        border-top: 6px solid transparent;
        border-bottom: 6px solid transparent;
        border-right: 6px solid var(--color-neutral-90);
        left: -6px;
        top: calc(50% - 6px);
        margin-left: 0;
        margin-right: 0;
    }
    [x-placement^="right-start"] &__arrow {
        top: var(--spacing-3x);
    }
    [x-placement^="right-end"] &__arrow {
        top: auto;
        bottom: var(--spacing-3x);
    }
    [x-placement^="left"] & {
        margin-right: var(--spacing-5x);
        transform: translateX(-20px);
    }
    [x-placement^="left"] &__arrow {
        border-top: 6px solid transparent;
        border-bottom: 6px solid transparent;
        border-left: 6px solid var(--color-neutral-90);
        right: -6px;
        top: calc(50% - 6px);
        margin-left: 0;
        margin-right: 0;
    }
    [x-placement^="left-start"] &__arrow {
        top: var(--spacing-3x);
    }
    [x-placement^="left-end"] &__arrow {
        top: auto;
        bottom: var(--spacing-3x);
    }
        
    @keyframes appearing {
        100% {
            opacity: 1;
            transform: translateX(0) translateY(0);
        }
    }
}
