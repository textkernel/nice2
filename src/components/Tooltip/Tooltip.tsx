import * as React from 'react';
import { PopupPlacement } from '../../constants';
import PopupBase from '../PopupBase';
import styles from './Tooltip.scss';
import bem from '../../utils/bem';

const { elem } = bem('Tooltip', styles);

interface Props {
    children: React.ReactElement | string;
    content: React.ReactElement | string;
    placement?: PopupPlacement;
}

const Tooltip: React.FC<Props> = props => {
    const { placement, content, children } = props;

    const createMouseEnterHandler = setPopupVisibility => () => {
        setPopupVisibility(true);
    };

    const createMouseLeaveHandler = setPopupVisibility => () => {
        setPopupVisibility(false);
    };

    const renderAnchor = ({ setPopupVisibility }) => (
        <div
            onMouseEnter={createMouseEnterHandler(setPopupVisibility)}
            onMouseLeave={createMouseLeaveHandler(setPopupVisibility)}
        >
            {children}
        </div>
    );

    const renderPopup = () => (
        <div>
            <div {...elem('container', props)}>
                {content}
                <div {...elem('arrow', props)} />
            </div>
        </div>
    );

    return (
        <PopupBase
            anchorRenderer={renderAnchor}
            popupRenderer={renderPopup}
            placement={placement}
        />
    );
};

Tooltip.displayName = 'Tooltip';

Tooltip.defaultProps = {
    placement: 'bottom',
};

export default Tooltip;
