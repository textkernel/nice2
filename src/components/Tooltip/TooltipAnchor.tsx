import * as React from 'react';
import { usePrevious } from '../../hooks';

interface Props extends React.HTMLAttributes<HTMLElement> {
    /** Anchor component */
    children: NotEmptySingleReactNode;
    /** indicates if there is a tooltip content or not */
    hasTooltipContent: boolean;
    /** If set to true will show the tooltip regardless of mouse position. When false, will show only on hover */
    alwaysVisible?: boolean;
    /** a function that toggles the visibility state of the tooltip */
    setPopupVisibility: (isOpen: boolean) => void;
    /** ref */
    ref?: React.RefObject<HTMLElement>;
}

export const TooltipAnchor: React.FC<Props> = React.forwardRef((props, ref) => {
    // eslint-disable-next-line react/prop-types
    const { hasTooltipContent, children, alwaysVisible, setPopupVisibility } = props;
    const prevAlwaysVisible = usePrevious(alwaysVisible);

    /** Behavior related to tooltip that is always open */
    React.useEffect(() => {
        if (alwaysVisible) {
            setPopupVisibility(hasTooltipContent);
        } else if (prevAlwaysVisible) {
            setPopupVisibility(false);
        }
    }, [hasTooltipContent, alwaysVisible, prevAlwaysVisible, setPopupVisibility]);

    if (alwaysVisible) {
        return React.isValidElement(children) ? (
            React.cloneElement(children, { ref })
        ) : (
            <span ref={ref}>{children}</span>
        );
    }

    /** Behavior related to tooltip that opens on hover only */
    const mouseOverHandler = () => {
        setPopupVisibility(true);
    };

    const mouseLeaveHandler = () => {
        setPopupVisibility(false);
    };

    const extraProps = {
        onMouseOver: mouseOverHandler,
        onMouseLeave: mouseLeaveHandler,
        ref,
    };

    if (React.isValidElement(children)) {
        return React.cloneElement(children, extraProps);
    }

    return <span {...extraProps}>{children}</span>;
});

TooltipAnchor.displayName = 'TooltipAnchor';

TooltipAnchor.defaultProps = {
    alwaysVisible: false,
    hasTooltipContent: false,
};
