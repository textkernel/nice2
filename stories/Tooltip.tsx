import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, select, text } from '@storybook/addon-knobs';
import { Tooltip } from '@textkernel/oneui';
import { POPUP_PLACEMENTS } from '../src/constants';

storiesOf('Molecules|Tooltip', module)
    .addDecorator(withKnobs)
    .add('Tooltip', () => (
        <div style={{ position: 'relative', display: 'table', margin: '0 auto' }}>
            <Tooltip
                placement={select('Placement', POPUP_PLACEMENTS, 'bottom')}
                content={text('Tooltip text', 'this is my tooltip text')}
            >
                hover here to see the tooltip
            </Tooltip>
        </div>
    ));
