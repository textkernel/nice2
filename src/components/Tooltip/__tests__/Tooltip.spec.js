import React from 'react';
import toJson from 'enzyme-to-json';
import Tooltip from '../Tooltip';

describe('<Tooltip> that renders a Tooltip', () => {
    it('should render default Tooltip correctly', () => {
        const wrapper = mount(
            <Tooltip placement="bottom" content="content">
                Hover me
            </Tooltip>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should display a tooltip on hover', () => {
        const wrapper = mount(
            <Tooltip placement="bottom" content="content">
                Hover me
            </Tooltip>
        );

        wrapper.childAt(0).simulate('mouseenter');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('div[data-popup="true"]')).toHaveLength(1);
    });
    it('should display proper text inside a tooltip', () => {
        const tooltipText = 'Tooltip text';
        const wrapper = mount(
            <Tooltip placement="bottom" content={tooltipText}>
                Hover me
            </Tooltip>
        );

        wrapper.childAt(0).simulate('mouseenter');
        expect(
            wrapper
                .childAt(0)
                .childAt(1)
                .text()
        ).toEqual(tooltipText);
    });
    it('should hide a tooltip on mouse leave', () => {
        const wrapper = mount(
            <Tooltip placement="bottom" content="content">
                Hover me
            </Tooltip>
        );

        wrapper.childAt(0).simulate('mouseenter');
        wrapper.childAt(0).simulate('mouseleave');
        expect(wrapper.find('div[data-popup="true"]')).toHaveLength(0);
    });
});
