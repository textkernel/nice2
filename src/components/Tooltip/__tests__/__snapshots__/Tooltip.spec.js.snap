// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip> that renders a Tooltip should display a tooltip on hover 1`] = `
<Tooltip
  content="content"
  placement="bottom"
>
  <PopupBase
    anchorRef={null}
    anchorRenderer={[Function]}
    placement="bottom"
    popupRef={null}
    popupRenderer={[Function]}
    renderInPortal={false}
  >
    <div
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Hover me
    </div>
    <div
      data-popup="true"
    >
      <div
        className="Tooltip__container"
      >
        content
        <div
          className="Tooltip__arrow"
        />
      </div>
    </div>
  </PopupBase>
</Tooltip>
`;

exports[`<Tooltip> that renders a Tooltip should render default Tooltip correctly 1`] = `
<Tooltip
  content="content"
  placement="bottom"
>
  <PopupBase
    anchorRef={null}
    anchorRenderer={[Function]}
    placement="bottom"
    popupRef={null}
    popupRenderer={[Function]}
    renderInPortal={false}
  >
    <div
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Hover me
    </div>
  </PopupBase>
</Tooltip>
`;
